### PR TITLE
Add support for secondary auth in IdentityCredentials

### PR DESCRIPTION
**Pull request by potto007**
~~WIP: exploring how to add Secondary Credentials support in the least obtrusive manner~~

### CODE REVIEW

**Comment by andrewpmartinez** [comment]
Can you elaborate on your use case? Are you trying to figure out how to authenticate w/ certificates and also provide a JWT? (I mention this because this is the only case I can figure that would need a primary auth w/ and a secondary).

**Comment by potto007** [comment]
> Can you elaborate on your use case? Are you trying to figure out how to authenticate w/ certificates and also provide a JWT? (I mention this because this is the only case that would need a primary auth w/ a secondary).

Hey @andrewpmartinez - yes, I am finishing up making authentication with primary and secondary work when secondary is an external jwt. I tested it against the Controller several weeks ago using Postman, and verified that the auth works when using a cert identity along with a JWT bearer on every request.

I realized I went a bit too far on this new type, and I'm pairing it down to just augment IdentityCredentials to include a JWT field.

**Comment by potto007** [comment]
Cleaning up, removing the custom type and instead overloading the IdentityCredentials type with optional JWT header support.

**Comment by potto007** [comment]
This update to sdk-golang now simply adds to the IdentityCredentials type. An example of its use can be seen in the zssh PR I made: https://github.com/openziti-test-kitchen/zssh/pull/31 

The existing functionality in Edge is such that, at the end of the `cert` auth method is processing, the code checks for an auth policy, and if the auth policy has external JWT as a secondary auth, it goes on to check for a token in the header, validates it, and succeeds if everything is in order.

**Review by andrewpmartinez — CHANGES_REQUESTED** [review]
I think this would be better as an improvement to `BaseCredentials` instead and de-JWT'ed. Rather add a `Headers map[string][]string{}`. Then augment the BaseCredentials.AuthenticateRequest to blindly apply those headers to the `runtime.ClientRequest`. Then alter all the other credentials first to call `self.BaseCredentials.AuthenticateRequest(cr, r)`.

The result is that credentials instantiations can do something like:


```
jwtStr := getJwt()
cred := NewIdentityCrednetialFromConfig(config)
cred.Headers["authorization"][]string{"Bearter " + jwtStr}
```

...or any other header that comes along. 

Reasoning:

1) Placing the functionality here limits the applicability to other cred mechanisms
2) Other authentication mechanisms are coming that are not JWTs but have header implications

File: ziti/contexts.go [hunk 1]
```diff
@@ -127,7 +128,16 @@ func NewContextWithOpts(cfg *Config, options *Options) (Context, error) {
 	}
 
 	if cfg.ID.Cert != "" && cfg.ID.Key != "" {
-		cfg.Credentials = edge_apis.NewIdentityCredentialsFromConfig(cfg.ID)
+		idCreds := edge_apis.NewIdentityCredentialsFromConfig(cfg.ID)
```
[the remark this answers is not in the document]
Comment (potto007) [reply ziti/contexts.go:131]: Gotcha - I had wondered why the implicit nature here - that background makes perfect sense.

**Comment by potto007** [comment]
That would be so much nicer! I'll give that a shot.

**Comment by potto007** [comment]
@andrewpmartinez I've refactored credentials.go based on your review.

**Comment by andrewpmartinez** [comment]
> 

Gonna look at this today.

File: edge-apis/credentials.go [hunk 2]
```diff
@@ -109,15 +124,43 @@ func (c *BaseCredentials) GetCaPool() *x509.CertPool {
 	return c.CaPool
 }
 
+// Method provides a base implementation to return the authentication method of a Credentials instance.
+func (c *BaseCredentials) Method() string {
```
Comment (andrewpmartinez) [inline edge-apis/credentials.go:128]: This method was purposefully not implemented as it stops its usage as a `Credentials` interface implementor. The BaseCredentials struct is meant to be embedded as a component of another struct that provides that method and fulfills the interface.

File: ziti/contexts.go [hunk 3]
```diff
@@ -87,7 +87,12 @@ func NewContextWithOpts(cfg *Config, options *Options) (Context, error) {
 	}
 
 	if cfg.ID.Cert != "" && cfg.ID.Key != "" {
+		baseCreds := &edge_apis.BaseCredentials{}
```
Comment (andrewpmartinez) [inline ziti/contexts.go:90]: This if-block should not change. It is for backward compatibility for systems that rely only on the previous file configuration logic.

If an implementor wants to use the new credentials interface, I would expect them to create their own credentials object and include any headers on that field. I don't want to mix the usage of the `cfg.ID` field and the `cfg.Credentials`.

**Review by andrewpmartinez — CHANGES_REQUESTED** [review]
See comments.

File: edge-apis/credentials.go [hunk 4]
```diff
@@ -71,34 +78,42 @@ func getClientAuthInfoOp(credentials Credentials, client *http.Client) func(*run
...
+	// SdkInfo is provided during authentication to set SDK information about the client.
 	SdkInfo *rest_model.SdkInfo
 
-	//CaPool will override the client's default certificate pool if set to a non-nil value.
+	// CaPool will override the client's default certificate pool if set to a non-nil value.
 	CaPool *x509.CertPool
 }
 
+// NewCredentials creates a new BaseCredentials instance.
+func NewCredentials() *BaseCredentials {
```
Comment (andrewpmartinez) [inline edge-apis/credentials.go:98]: This should not exist.

File: example/zping/go.work.sum [hunk 5]
```diff
@@ -33,6 +33,7 @@ github.com/cyberdelia/templates v0.0.0-20141128023046-ca7fffd4298c/go.mod h1:GyV
 github.com/deepmap/oapi-codegen v1.3.13/go.mod h1:WAmG5dWY8/PYHt4vKxlt90NsbHMAOCiteYKZMiIRfOo=
```
Comment (andrewpmartinez) [inline example/zping/go.work.sum:33]: This file should not be included in the pr.

File: ziti/contexts.go (hunk 3 above)
Comment (potto007) [reply ziti/contexts.go:90]: I modified this conditional in an attempt to avoid side-effects. The problem is that BaseCredentials gets overwritten by the `NewIdentityCredentialsFromConfig` call.

File: edge-apis/credentials.go (hunk 4 above)
Comment (potto007) [reply edge-apis/credentials.go:98]: This is how I was using it: https://github.com/openziti-test-kitchen/zssh/blob/37c5561e1747ea312584487ca6ede3502702713c/zsshlib/ssh.go#L329

The problem is that, short of that, one needs to create another Credentials type that doesn't match the intent, making it less expressive... for instance, I could change the accessor to perform the following:

```
conf := getConfig(f.ZConfig)
conf.Credentials = edge_apis.NewJwtCredentials("")
conf.Credentials.AddHeader("Authorization", "Bearer "+token)
```
which would work, but gives the misconception that the JwtCredential type is actually being leveraged or is necessary.

File: edge-apis/credentials.go (hunk 2 above)
Comment (potto007) [reply edge-apis/credentials.go]: I've removed it. It doesn't need to be an interface so long as the SDK user knows to update credentials after the `NewContext` call.

File: ziti/contexts.go (hunk 3 above)
Comment (potto007) [reply ziti/contexts.go]: I've reverted this to the original conditional. The issue this conditional presents is that it creates a side-effect on the Credentials. My update to the code was an attempt to allow for the existing workaround behavior without it causing strange behavior. 

The workaround is to ensure that the SDK docs make it explicit that the credentials are not stable until _after_ the NewContext call is made.

Comment (potto007) [reply ziti/contexts.go]: An example of what I'm talking about is here: https://github.com/openziti-test-kitchen/zssh/blob/d6425563e497e5f93b2225816cc4a2a753655aab/zsshlib/ssh.go#L329 -- I needed to move the update of Credentials to be _after_ the call to `NewContext` in order to avoid `NewContextWithOpts` from overwriting `BaseCredentials.Headers`

**Comment by potto007** [comment]
@andrewpmartinez any thoughts?

**Review by andrewpmartinez — APPROVED** [review]

File: edge-apis/credentials.go [hunk 6]
```diff
@@ -8,26 +8,33 @@ import (
...
-	//Method return the authentication necessary to complete an authentication request.
+	// Method returns the authentication necessary to complete an authentication request.
 	Method() string
 
-	//ClientAuthInfoWriter is used to pass a Credentials instance to the openapi runtime to authenticate outgoing
+	// AddHeader adds a header to the request.
+	AddHeader(key, value string)
+
+	// AuthenticateRequest authenticates an outgoing request.
+	AuthenticateRequest(runtime.ClientRequest, strfmt.Registry) error
```
[the remark this answers is not in the document]
Comment (potto007) [reply edge-apis/credentials.go:35]: Good point - suggest I rename?
